docs: Add ADR-0017 Time-Bound Quality Ladder for temporal asset reconciliation

bjcoombs opened the pull request.

claude committed:
docs: Add ADR-0017 Time-Bound Quality Ladder for temporal asset reconciliation

Introduces architectural pattern for handling competing data quality over time
periods, particularly relevant for metered assets like energy consumption.

Key concepts:
- Universal Time Model with Period [Start, End] ranges
- Source Authority Registry for quality score derivation
- Append-only Measurement Log with supersession tracking
- Delta Engine for evaluating incoming measurements
- Wash & Reload correction saga (reversal + booking, no mutations)
- Bill binding for reconciliation and dispute handling
- Settlement finality windows (D+1, D+5, M+14)

Status: Proposed

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/meridianhub/meridian?pullRequest=305) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/meridianhub/meridian?pullRequest=305) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/meridianhub/meridian?pullRequest=305) it.</sub>

claude committed:
docs: Address review feedback on ADR-0017

Improvements:
- Add concurrency documentation with tiebreaker for simultaneous arrivals
- Add Error Taxonomy section with domain-specific errors
- Fully define EntryType enum (BOOKING, CORRECTION_REVERSAL, CORRECTION_BOOKING,
  TRANSFER, ADJUSTMENT, DISPUTE)
- Document overlap prevention enforcement with SERIALIZABLE isolation
- Add bill binding performance notes with pagination for high-volume bills
- Add backfill window configuration schema (asset_settlement_rules table)
- Standardize ID types: AccountID as uuid.UUID throughout
- Clarify Bill Binding service responsibility (creation vs storage)
- Add Scope and Boundaries section clarifying:
  - Dispute workflow is out of scope (target ADR-0018)
  - Valuation engine is out of scope (target ADR-0019)
  - Attribute handling semantics (JSONB containment)

claude committed:
docs: Address second round of review feedback on ADR-0017

Changes addressing technical requirements:
- Replace SERIALIZABLE isolation with optimistic concurrency via position key
  hash for 100k TPS throughput requirement
- Fix Bill Binding → Settlement Snapshot (Financial Accounting owns its own
  schema, no cross-schema queries)
- Add Considered Options section (event sourcing, bi-temporal, etc.)
- Add dispute fail-safe behavior (never silently drop data)
- Add settlement run validation (prevent free-form strings)

Terminology updates:
- Replace all "bill/billing" with "settlement" terminology - this is an
  accounting/treasury system, not a billing engine

Domain-specific improvements:
- Add compute time example (COMPUTE_HOURS asset)
- Update Valuation Integration to include cross-asset conversion
- Make examples more generic (units instead of kWh)
- Reference compareSameQuality in Delta Engine for tiebreaking

Project rules addressed:
- Immutable: maintained (append-only, Wash & Reload)
- No cross-schema queries: Settlement Snapshots stay in Financial Accounting
- 100k TPS: Position Key Hash with unique constraint, no locks

claude committed:
docs: Address third round of review feedback on ADR-0017

Technical fixes:
- Period.Overlaps(): Document half-open interval semantics [Start, End)
  and handle instant events correctly
- Remove duplicate Supersede() function, add retry strategy documentation
  with IngestWithRetry() example
- Fix JSONB hash determinism: Add canonicalize_jsonb() function for
  sorted key ordering in position_key_hash
- Add bounds validation to settlement run (D+0..30, M+1..24)
- Document reconciliation fail-fast error handling strategy with rationale

Operational concerns addressed:
- Hash collision: Document probability (~0) and operational impact
- Data retention: Add tiered storage strategy (hot/warm/cold/archive)
- Snapshot denormalization: Justify 560GB/year tradeoff for query locality
  and cross-service isolation

Added sections:
- Data Retention and Archival (with example archival SQL)
- Settlement Snapshot Denormalization Justification
- Reconciliation Error Handling comparison table
- Retry Strategy for ErrAlreadySuperseded

claude committed:
docs: Address final review feedback on ADR-0017

Clarifications:
- Period semantics: Add table explaining why Overlaps() uses half-open [Start, End)
  while Contains() uses closed [Start, End] intervals
- TSTZRANGE handling: Document PostgreSQL instant handling with closed-closed bounds

Missing pieces:
- Add idx_measurements_settlement_run index for FinalizeRun queries
- Add PositionEntry.GetAttributes() method with note that attributes are derived
  from MeasurementID (avoids duplication)
- Add ValuationEngine placeholder interface (ValuationRequest, ValuationResult)
- Add ErrOutsideBackfillWindow validation in Ingest() function

Design Decisions FAQ:
- Attribute subset matching: No - exact equality required, use GIN index for analytics
- Configurable settlement run bounds: Deferred - defaults cover all known conventions
- Backfill enforcement location: MeasurementIngestionService.Ingest()

ADR now at ~1400 lines with comprehensive coverage.

claude committed:
docs: Address continued review feedback on ADR-0017

- Fix Overlaps() for two instants at same timestamp returning false
- Add dedicated Hash Collision Risk Assessment subsection with table
- Refactor Reconcile() to use BatchFindCurrent (N+1 query fix)
- Add GIN index for attribute-based queries
- Document dispute event transaction boundary with outbox pattern
- Add FAQ answers for attribute mutability, multi-period corrections,
  and Source Authority retroactive versioning

claude committed:
docs: Clarify quality score semantics and add Attributes to snapshots

- Clarify that Source Authority Registry is the canonical source for
  quality rankings, with measurements carrying optional denormalized
  snapshots that may diverge post-ingestion
- Add mirror clarification to Measurement.QualityScore field comment
- Add Attributes field to SettlementSnapshot for fungibility context
- Define PositionKey struct with Attributes per ADR-0014 requirements
- Update Reconcile() to include Attributes in position key lookups
- Update BatchFindCurrent() to document attribute-aware matching

claude committed:
docs: Add tenant isolation, security, observability to ADR-0017

Critical fixes:
- Add tenant_id to measurements table with isolation index (ADR-0016)
- Add UTC validation to Period constructor and Validate()
- Fix duplicate detection to check period + attributes, not just source + quantity
- Add ActionError sentinel (-1) to avoid misleading 0 return on errors
- Add context propagation to CorrectionSaga via WithContext()

New sections:
- Event Contracts: MeasurementSuperseded, SettlementRunCompleted, etc. (ADR-0004)
- Security Considerations: Settlement locking authorization, dispute rate limiting
- Observability: Metrics, tracing, alerting thresholds
- Glossary: Quality Ladder, Wash & Reload, Settlement Run, etc.
- TSTZRANGE Integration Tests: Required test cases for instant handling

Updated Links to reference ADR-0004 and ADR-0016.

claude committed:
docs: Fix Overlaps() to use closed-interval semantics per ADR

Critical fix: The ADR documented "[Start, End] ranges" (closed intervals)
but Overlaps() implemented half-open [Start, End) semantics. This caused:
- Instant Period{t, t} would not overlap itself under half-open logic
- Contradicted Contains() which correctly used closed semantics
- Misaligned with documented position model

Changes:
- Simplify Overlaps() to single closed-interval test:
  !p.Start.After(other.End) && !other.Start.After(p.End)
- Update semantics table to show both methods use closed [Start, End]
- Add note about PostgreSQL TSTZRANGE translation for closed bounds
- Add ADR-0017 to Data Management category in README.md

claude committed:
docs: Add Temporal Authority and Provisional Settlement to ADR-0017

This transforms the ADR from "a ledger that fixes errors" to "a ledger
that manages time" - enabling forward-looking capabilities like hedging,
budgeting, and resource reservation.

Key additions:

1. Source Authority with Temporal Context:
   - Added FORECAST_ML, FORECAST_BUDGET sources
   - New TemporalContext field: "PAST", "FUTURE", or "ANY"
   - Rules: Past sources compete for historical periods,
     Future sources (forecasts) are authoritative for T+1 onwards

2. Section 6: Temporal Authority & Provisional Settlement:
   - "Freezing" mechanism preserves decision provenance
   - ProvisionalSettlement creates snapshots with forecast basis
   - ReconcileProvisional generates FORECAST_DEVIATION variances
   - Cross-domain applications table (Energy/NGO/Compute/Treasury)
   - NGO Budget Allocation example with full lifecycle

3. SettlementSnapshot provenance fields:
   - SourceAtSettle: The source used (e.g., "FORECAST_ML")
   - SettlementType: PROVISIONAL or FINAL
   - Enables "why did we settle that amount?" audit trail

4. Glossary updates:
   - Temporal Authority, Provisional Settlement, Freezing, ATP

Strategic value: This is the specific mechanic that makes "Murex"
(forward-looking) work with "10x" (backward-looking) capabilities.

claude committed:
docs: Add race condition handling and defensive tests to ADR-0017

Addresses review feedback:

1. Consequences > Positive:
   - Added "Unified Treasury & Accounting" - the key business value that
     enables cashflow forecasting and historical audit in a single view

2. Concurrency Handling - High-Contention Position Locking:
   - Added distributed lock implementation guidance for high-frequency
     positions (>100 writes/sec)
   - Added decision table: when to use optimistic retry vs distributed lock
   - Covers normal ingestion, high-frequency aggregation, batch imports,
     and settlement run scenarios

3. Defensive Tests (Per ADR-0008):
   - Added TestCorrectionSaga_Execute_DefensiveTests with 4 scenarios:
     * Old measurement already superseded (double-supersession)
     * Cross-account corruption check
     * Transaction timeout with rollback verification
     * Locked position rejection
   - Each test verifies no partial state on failure

Note: GIN index for attributes already present from prior commit.
Note: ActionError sentinel retained - provides self-documenting code
      and prevents accidental use of Action when error != nil.

claude committed:
docs: Split ADR-0017 into Physics and Lifecycle ADRs

Split the cognitively overwhelming ADR-0017 into two focused documents:

- ADR-0017 (Temporal Quality Ledger - Data Physics): Measurement log,
  quality ladder, supersession, Delta Engine, Wash & Reload correction

- ADR-0018 (Settlement & Reconciliation - Lifecycle): Temporal authority,
  settlement snapshots, reconciliation workflows, dispute handling

This separation allows database engineers to focus on storage mechanics
while domain engineers focus on settlement business logic.

bjcoombs commented:
Superseded by #310 which consolidates ADR-0017 and ADR-0018 into a clean PR from develop.
